synapse/storage/databases/main/events.py

```diff
@@ -410,6 +410,31 @@ def _persist_events_txn(
         assert min_stream_order
         assert max_stream_order
 
+        # Once the txn completes, invalidate all of the relevant caches. Note that we do this
+        # up here because it captures all the events_and_contexts before any are removed.
+        for event, _ in events_and_contexts:
+            self.store.invalidate_get_event_cache_after_txn(txn, event.event_id)
+            if event.redacts:
+                self.store.invalidate_get_event_cache_after_txn(txn, event.redacts)
+
+            relates_to = None
+            relation = relation_from_event(event)
+            if relation:
+                relates_to = relation.parent_id
+
+            assert event.internal_metadata.stream_ordering is not None
+            txn.call_after(
+                self.store._invalidate_caches_for_event,
+                event.internal_metadata.stream_ordering,
+                event.event_id,
+                event.room_id,
+                event.type,
+                getattr(event, "state_key", None),
+                event.redacts,
+                relates_to,
+                backfilled=False,
+            )
```

Comment:
For my own understanding, how does this get around the `ValueError: Cache objects can only be accessed from the main thread` problem? I ran into this while calling `_invalidate_caches_for_event` directly here, https://github.com/matrix-org/synapse/pull/13863#discussion_r977056388

I assume `txn.call_after(...)` is doing the magic here but it's not clear to me how since `_persist_events_txn` is still in the event persistence queue layer being processed in the background.

Reply:
I believe the issue is that DB txns are executed on a separate thread (twisted's [workaround for DBAPI compatibility](https://twisted.org/documents/10.1.0/core/howto/rdbms.html)). The `call_after` callbacks are executed back on the main thread after the txn completes (https://github.com/matrix-org/synapse/blob/1a1abdda42551dad3aadc04a169c25f4cc651a2c/synapse/storage/database.py#L866).

Reply:
Ahhh, so the problem isn't the event persistence queue, it's that we're within a database transaction (`db_pool.runInteraction("persist_events", self._persist_events_txn, ...)`)

Thanks for the info and links! 🙂